File: day3/map.go
package day3

import (
	"math"
)

type point struct {
	X int
	Y int
}

func (p point) distance() int {
	return abs(p.X) + abs(p.Y)
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

type cell struct {
	id    int8
	steps int
}

type grid map[point]map[int8]cell

type Map struct {
	grid grid
}

func NewMap() *Map {
	return &Map{
		grid: make(grid),
	}
}

func (m *Map) Count(x, y int) int {
	return m.countAtPoint(point{x, y})
}

func (m *Map) countAtPoint(p point) int {
	return len(m.grid[p])
}

func (m *Map) Set(x, y int, id int8, steps int) {
	p := point{x, y}
	cells := m.grid[p]
	if cells == nil {
		cells = make(map[int8]cell)
		m.grid[p] = cells
	}

	cell, ok := cells[id]
	if !ok {
		cell.id = id
		cell.steps = steps
		cells[id] = cell
	}
}

func (m *Map) NearestIntersection() (int, int, int) {
	var best point
	d := math.MaxInt64
	for p := range m.grid {
		n := m.countAtPoint(p)
		if n < 2 {
			continue
		}

		if best.X == 0 && best.Y == 0 {
			best = p
			d = p.distance()
		} else {
			nd := p.distance()
			if nd < d {
				best = p
				d = nd
			}
		}
	}

	return best.X, best.Y, d
}

func (m *Map) ShortestIntersection() (int, int, int) {
	var best point
	steps := math.MaxInt64
	for p, cells := range m.grid {
		if len(cells) < 2 {
			continue
		}

		var totalSteps int
		for _, cell := range cells {
			totalSteps += cell.steps
		}

		if best.X == 0 && best.Y == 0 {
			best = p
			steps = totalSteps
		} else {
			if totalSteps < steps {
				best = p
				steps = totalSteps
			}
		}
	}

	return best.X, best.Y, steps
}

File: aoc2019_elixir/apps/day03/lib/map.ex
defmodule Day03.Map do
  @moduledoc """
  Functions for working with wiring maps.
  """

  @typedoc """
  A map of points that have been visited by a path of wires.
  """
  @type t :: map

  @typedoc """
  A point on the map in `{x, y}` coordinates.
  """
  @type point :: {integer, integer}

  @typedoc """
  An identifier for a particular wire path.

  This is used to identify which points on the map have the paths crossing.
  """
  @type id :: atom

  @doc """
  Sets the number of steps required to reach a point on the map along a given path.
  """
  @spec set(t, point, id, integer) :: t
  def set(m, point, id, steps) do
    Map.update(m, point, %{id => steps}, &Map.put_new(&1, id, steps))
  end

  @doc """
  Returns the point of the nearest intersection of two wires.

  The nearest point is determined by Manhattan distance from the origin `{0, 0}`.
  """
  @spec nearest_intersection(t) :: {point, integer}
  def nearest_intersection(m) do
    Enum.filter(m, fn {_, cells} -> Enum.count(cells) >= 2 end)
    |> Enum.map(fn {point, _} -> {point, point_distance(point)} end)
    |> Enum.min_by(fn {_, distance} -> distance end)
  end

  @doc """
  Returns the intersection of two wires that requires the fewest steps.

  This requires following the actual paths to know how many steps along each path
  it takes to get to the point.
  """
  @spec shortest_intersection(t) :: {point, integer}
  def shortest_intersection(m) do
    Enum.filter(m, fn {_, cells} -> Enum.count(cells) >= 2 end)
    |> Enum.map(fn {point, cells} ->
      {point, Enum.reduce(cells, 0, fn {_, steps}, acc -> acc + steps end)}
    end)
    |> Enum.min_by(fn {_, steps} -> steps end)
  end

  defp point_distance({x, y}) do
    abs(x) + abs(y)
  end
end
